### DISCUSSION TITLE
Pixel Perfect Games & Jitter

### DISCUSSION THREAD

**Discussion by Phanterm** [ramokz/phantom-camera#180] [Question / Help Wanted]
Hi there,

First off, this plugin is wonderful, and I can't wait to begin utilizing it to the fullest. But that's also where my problem lies. I know how to set up Phantom Camera, but what are the best practices for doing so? I've made two different implementations with different problems for each.

For both test implementations, the following properties are set:
- `pixel_perfect` set to `true`
- `follow_mode` set to Framed
- `limit_node_target` is assigned to a large CollisionShape2D for the test room

### TYPE A: Autoloader/Remote Camera

https://github.com/ramokz/phantom-camera/assets/3878045/3bfab161-4cb3-49ed-b83f-c4fa2ff47c3c

In the video I attached, the camera used here exists on an Autoloader named CameraManager. The node setup is as follows: 

![image](https://github.com/ramokz/phantom-camera/assets/3878045/9400d096-a652-4fc9-ac2b-2c2bd04ddbf1)

As you can see, the player node is set to be the follow target. However, even with `pixel_perfect` enabled, I experience jitter when the camera moves. 

### TYPE B: Child Camera

https://github.com/ramokz/phantom-camera/assets/3878045/0138c42b-f1d1-4730-8974-93c872fa1b2a

The camera lives as a child of the Player node. 

![image](https://github.com/ramokz/phantom-camera/assets/3878045/3d5a8bbc-3f88-403d-b41b-6e93307d51bf)

Whether or not I set the follow target is irrelevant, as the camera moves with the player anyway as a child object. 

### ISSUES

The Autoloader solution is best for controlling my cameras from a central location, and that's ultimately easier for me. However, the pixel perfect property isn't enough to offset the strange jitter I'm running into. I've been having this issue with Godot's regular cameras for some time and haven't found a solution after a lot of research.

The child camera is a direct solution that, paired with pixel perfect, gives me that smooth motion. The downside is that it basically forces me into a glued follow mode, cannot be controlled globally like the Autoloader, and is generally more cumbersome to deal with.

I suspect the problem lies outside of PhantomCamera when it comes to the jitter issue, but given that I've exhausted just about every other help forum when it comes to cameras, I figure I'd ask here if I missed anything.

Thank you again!

**Comment by Phanterm** [comment]
Closing because I was experimenting a bit more since this, and disabled v-sync in my project settings, which eliminated 90% of the jitter. Apologies for cluttering the discussion board!

**Reply by ramokz** [reply]
Not at all! Think there's still work to be done on the jittery side, for pixel perfection especially.

Haven't tried overly many ways of putting `PCams` together in terms of scene structures. But based on your video demonstrations, it does appear to work well. There isn't really a “best practice” per se, if it works for you then that's a valid approach. The only thing an ideal scenario would be is to have both the `Camera2D` and `PCams` in the same scene to more easily see how changes affect the final framing, but that's more of a developer experience perk.

By the looks of it, I'm willing to suspect the jitter is due to either the way the camera interpolates (from the addon) or a `Framed Follow` quirk. Either way, are currently looking at camera jittery in #179 and #161 is a refactor for `Framed Follow`, which might also help. To the former PR, it's also worth checking out the [smoothing-addon](https://github.com/lawnjelly/smoothing-addon/tree/4.x) combined with the changes in the #179 and see if that helps with anything.

> Whether or not I set the follow target is irrelevant, as the camera moves with the player anyway as a child object.

This is a bug I've seen before. The refactor for `Framed Follow` should hopefully address that.
